app/models/user.rb

```diff
@@ -17,24 +17,47 @@ class User < ApplicationRecord
          :recoverable, :rememberable, :validatable,
          :doorkeeper, :omniauthable, omniauth_providers: [:google_oauth2]
 
+  enum provider: { email: 'email', google_oauth2: 'google_oauth2' }
+
+  validates :provider, inclusion: { in: providers.keys }
+
   class << self
-    def from_omniauth(auth)
-      User.find_or_create_by(provider: auth.provider, uid: auth.uid, email: auth.info.email) do |user|
-        set_user_attributes(user, auth)
-      end
+    def from_email(email, password)
+      user = find_or_initialize_user(
+        email: email,
+        password: password,
+        provider: User.providers[:email]
+      )
+
+      user.persisted? ? user.errors.add(:email, :taken) : user.save
```

Comment:
If the `email` field is unique, I think we don't need this line, as it will be handled by the User validation 🤔

Reply:
Since we're using `find_or_initialize_user`, we need `user.persisted?` to know if the record already existed or not.

Without this line, it will just simply return the existing user, without giving any error

Reply:
The concern raised by Hoang comes from the fact that team rarely uses `find_or_initialize_user` 😅 Also, you have not revisited this implementation based on an earlier comment about the ternary operator not being idiomatic in Ruby

Reply:
Thank you guys, I was indeed overcomplicating it a lot 🙈 

I have simplified the code, looks a lot better now  🚀